shared/clientstats/scrapers.go

```diff
@@ -0,0 +1,197 @@
+package clientstats
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prom2json"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+func scrapeProm(url string) (map[string]*dto.MetricFamily, error) {
+	mfChan := make(chan *dto.MetricFamily, 1024)
+	err := prom2json.FetchMetricFamilies(url, mfChan, http.DefaultTransport)
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]*dto.MetricFamily)
+	for fam := range mfChan {
+		ptr := fam
+		result[fam.GetName()] = ptr
+	}
+	return result, nil
+}
+
+type validatorScraper struct {
+	url string
+}
+
+func (vc *validatorScraper) Scrape() (io.Reader, error) {
+	log.Infof("Scraping validator at %s", vc.url)
+	pf, err := scrapeProm(vc.url)
+	if err != nil {
+		return nil, nil
+	}
+
+	cs, err := populateCommonStats(pf)
+	if err != nil {
+		return nil, err
+	}
+	vs, err := populateValidatorStats(pf)
+	if err != nil {
+		return nil, err
+	}
+	vs.CommonStats = cs
+
+	b, err := json.Marshal(vs)
+	return bytes.NewBuffer(b), err
+}
+
+func NewValidatorScraper(promExpoURL string) Scraper {
+	return &validatorScraper{
+		url: promExpoURL,
+	}
+}
+
+type beaconNodeScraper struct {
+	url string
+}
+
+func populateCommonStats(pf map[string]*dto.MetricFamily) (CommonStats, error) {
+	cs := CommonStats{}
+	cs.ClientName = "prysm"
+	var f *dto.MetricFamily
+	var m *dto.Metric
+	var ok bool
+
+	f, ok = pf["process_cpu_seconds_total"]
+	if ok {
+		m = f.Metric[0]
+		// float64->int64: truncates fractional seconds
+		cs.CPUProcessSecondsTotal = int64(m.Counter.GetValue())
+	}
+
+	f, ok = pf["process_resident_memory_bytes"]
+	if ok {
+		m = f.Metric[0]
+		cs.MemoryProcessBytes = int64(m.Gauge.GetValue())
+	}
+
+	f, ok = pf["prysm_version"]
+	if ok {
+		m = f.Metric[0]
+		for _, l := range m.GetLabel(){
+			switch l.GetName() {
+			case "version":
+				cs.ClientVersion = l.GetValue()
+			case "buildDate":
+				buildDate, err := strconv.Atoi(l.GetValue())
+				if err != nil {
+					return cs, fmt.Errorf("Error when retrieving buildDate label from the prysm_version metric: %s", err)
+				}
+				cs.ClientBuild = int64(buildDate)
+			}
+		}
+	}
+
+	return cs, nil
+}
+
+func populateValidatorStats(pf map[string]*dto.MetricFamily) (ValidatorStats, error) {
+	vs := ValidatorStats{}
+
+	//var f *dto.MetricFamily
+
+	/*
```

Comment:
Let's avoid checking-in commented code

Reply:
juuuust keeping this around because I have a feeling that we might move it over to beacon-node. Will remove before we merge for sure.